apps/api/src/auth/oslo-auth.ts

```diff
@@ -0,0 +1,103 @@
+import { sha256 } from "@oslojs/crypto/sha2";
+import { encodeBase32LowerCaseNoPadding, encodeHexLowerCase } from "@oslojs/encoding";
+import { eq } from "drizzle-orm";
+import type { Context } from "hono";
+import { env } from "hono/adapter";
+
+import type { AppContext } from "./../context";
+import { sessionTable } from "./../database/sessions";
+import { userTable } from "./../database/users";
+import type { DatabaseUserAttributes, Session, SessionValidationResult } from "./../types";
+
+type SessionQueryResult = {
+  user: DatabaseUserAttributes & { id: string };
+  session: {
+    id: string;
+    userId: string;
+    expiresAt: number;
+  };
+};
+
+export function generateSessionToken(): string {
+  const bytes = new Uint8Array(20);
+  crypto.getRandomValues(bytes);
+  return encodeBase32LowerCaseNoPadding(bytes);
+}
+
+export async function createSession(c: Context<AppContext>, token: string, userId: string): Promise<Session> {
+  const db = c.get("db");
+  const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));
+  const session: Session = {
+    id: sessionId,
+    userId,
+    expiresAt: 60 * 60 * 24 * 30,
+  };
+
+  await db.insert(sessionTable).values(session);
+  return session;
+}
+
+export async function validateSessionToken(c: Context<AppContext>, token: string): Promise<SessionValidationResult> {
+  const db = c.get("db");
+  const sessionId = encodeHexLowerCase(sha256(new TextEncoder().encode(token)));
+
+  const results = await db
+    .select({
+      session: sessionTable,
+      user: userTable,
+    })
+    .from(sessionTable)
+    .where(eq(sessionTable.id, sessionId))
+    .innerJoin(userTable, eq(sessionTable.userId, userTable.id));
+
+  const result = results[0] as SessionQueryResult | undefined;
+  if (!result) {
+    return { session: null, user: null };
+  }
+
+  const { user, session } = result;
+
+  if (Date.now() >= Date.now() + session.expiresAt) {
+    await db.delete(sessionTable).where(eq(sessionTable.id, session.id));
+    return { session: null, user: null };
+  }
```

Comment:
As Coderabbit suggested, this line is like

1 >= 1+ 232(let it be anything)
it will always be false
```Date.now() >= Date.now() - session.expiresAt```
this should be it.

Reply:
yeah sorry such a silly mistake.